File: tests/perl5/test-RNA-ensemble_defect.pl
#!/usr/bin/perl -I.libs

######################### We start with some black magic to print on failure.
# (It may become useful if the test is moved to ./t subdirectory.)
use strict;
use Test::More tests => 6;
use Data::Dumper;
use FileHandle;

use RNA;
use warnings;


######################### End of black magic.

##########################################################################################################################
#starting with ensemble_defect test

my $seq       = "AGGAAACCUUAAUUGGUUA";
my $structpk  = ".((...))(([[..))]].";
my $seq2      = "AAAAAAAA";
my $struct2   = "(......)";
my $struct3   = "..(..)..";

####################################################
##test_ensemble_defect
print "test_ensemble_defect\n";
my $fc = new RNA::fold_compound($seq);
(my $ss, my $gfe) = $fc->pf();

my $ed = $fc->ensemble_defect($structpk);
is($ed, 0.614080983833787, "Ensemble Defect");

my $pt = RNA::ptable($structpk);
$ed = $fc->ensemble_defect($pt);
is($ed, 0.614080983833787, "Ensemble Defect (pair table)");

$ed = $fc->ensemble_defect($structpk, RNA::BRACKETS_ANY);
is($ed, 0.7279184335061499, "Ensemble Defect (pk)");

$pt = RNA::ptable($structpk, RNA::BRACKETS_ANY);
$ed = $fc->ensemble_defect($pt);
is($ed, 0.7279184335061499, "Ensemble Defect (pk, pair table)");


$fc = new RNA::fold_compound($seq2);
($ss, $gfe) = $fc->pf();

$ed = $fc->ensemble_defect($struct2) * length($seq2);
is($ed, 2.0, "Ensemble Defect (Sanity check 1)");

$ed = $fc->ensemble_defect($struct3) * length($seq2);
is($ed, 2.0, "Ensemble Defect (Sanity check 2)");

undef $pt;
undef $fc;

File: tests/python/test_RNA-ensemble_defect.py
import unittest


if __name__ == '__main__':
    from py_include import taprunner
    import RNApath
    RNApath.addSwigInterfacePath()

import RNA


seq          = "AGGAAACCUUAAUUGGUUA"
structpk     = ".((...))(([[..))]]."
seq2         = "AAAAAAAA"
struct2      = "(......)"
struct3      = "..(..).."

class ensemble_defectTest(unittest.TestCase):
    def test_ensemble_defect_sanity(self):
        """Ensembe defect (sanity check)"""
        fc = RNA.fold_compound(seq2)
        fc.pf()

        ed = fc.ensemble_defect(struct2) * len(seq2)
        self.assertEqual(ed, 2.0)

        ed = fc.ensemble_defect(struct3) * len(seq2)
        self.assertEqual(ed, 2.0)

    def test_ensemble_defect(self):
        """Ensembe defect"""
        fc = RNA.fold_compound(seq)
        fc.pf()

        ed = fc.ensemble_defect(structpk)
        self.assertEqual(ed, 0.614080983833787)

        pt = RNA.ptable(structpk)
        ed = fc.ensemble_defect(pt)
        self.assertEqual(ed, 0.614080983833787)

    def test_ensemble_defect_pk(self):
        """Ensemble defect (pk)"""
        fc = RNA.fold_compound(seq)
        fc.pf()

        ed = fc.ensemble_defect(structpk, RNA.BRACKETS_ANY)
        self.assertEqual(ed, 0.7279184335061499)

        pt = RNA.ptable(structpk, RNA.BRACKETS_ANY)
        ed = fc.ensemble_defect(pt)
        self.assertEqual(ed, 0.7279184335061499)


if __name__ == '__main__':
    unittest.main(testRunner=taprunner.TAPTestRunner())
